File: src/main/java/kingdoms/server/handlers/packets/server/SPacketBuild.java
package kingdoms.server.handlers.packets.server;

import kingdoms.api.network.AbstractPacket;
import kingdoms.server.WorldProvider;
import kingdoms.server.handlers.Buildings;
import kingdoms.server.handlers.schematic.Schematic;
import kingdoms.server.handlers.schematic.SchematicHandler;
import kingdoms.server.handlers.schematic.SchematicRegistry;
import net.minecraft.entity.player.EntityPlayerMP;

public final class SPacketBuild extends AbstractPacket
{
    public SPacketBuild(){}
    public SPacketBuild(byte id, boolean guild)
    {
        buf().writeByte(id);
        buf().writeBoolean(guild);
    }

    @Override
    public void server(EntityPlayerMP player)
    {
        int id = buf().readByte();

        WorldProvider worldProvider = WorldProvider.get(player.worldObj);

        if (buf().readBoolean())
        {
            Schematic schematic = (new Schematic("/schematics/GuildCastle")).setPosition((int) player.posX, (int) player.posY, (int) player.posZ).setSpeed(300);
            SchematicHandler.INSTANCE.addBuilding(schematic);
            Buildings.INSTANCE.setBuildingTrue(0);
        }
        else {
            SchematicRegistry registry = SchematicRegistry.INSTANCE;

            if (!registry.isRegistered())
            {
                registry.registerAllBuildings((int) player.posX, (int) player.posY, (int) player.posZ);

                worldProvider.guildPosX = (int) player.posX;
                worldProvider.guildPosY = (int) player.posY;
                worldProvider.guildPosZ = (int) player.posZ;
                worldProvider.markDirty();
            }

            Schematic schematic = registry.getSchematic(id).setSpeed(300);
            SchematicHandler.INSTANCE.addBuilding(schematic);
            Buildings.INSTANCE.setBuildingTrue(id);
        }
    }
}

File: src/main/java/kingdoms/server/handlers/packets/server/SPacketBuild.kt
package kingdoms.server.handlers.packets.server

import kingdoms.api.network.SimplePacket
import kingdoms.server.TaleOfKingdoms
import kingdoms.server.WorldProvider
import kingdoms.server.handlers.Buildings
import kingdoms.server.handlers.schematic.Schematic
import kingdoms.server.handlers.schematic.SchematicHandler
import kingdoms.server.handlers.schematic.SchematicRegistry
import net.minecraft.entity.player.EntityPlayerMP
import net.minecraft.server.MinecraftServer
import net.minecraft.util.ChatComponentText

class SPacketBuild(): SimplePacket()
{
    constructor(id: Byte, guild: Boolean): this()
    {
        buf().writeByte(id.toInt())
        buf().writeBoolean(guild)
    }

    override fun server(player: EntityPlayerMP)
    {
        val id = buf().readByte().toInt()
        val guild = buf().readBoolean()

        if (TaleOfKingdoms.proxy.config.isOp())
        {
            if (MinecraftServer.getServer().configurationManager.func_152596_g(player.gameProfile))
            {
                build(id, guild, player)
            }
            else
            {
                player.addChatMessage(ChatComponentText("Sorry, you can\'t not building on this server!"))
            }
        }
        else
        {
            build(id, guild, player)
        }
    }

    private fun build(id: Int, guild: Boolean, player: EntityPlayerMP)
    {
        val worldProvider = WorldProvider.get(player.worldObj)

        val posX: Int = player.posX.toInt()
        val posY: Int = player.posY.toInt()
        val posZ: Int = player.posZ.toInt()

        if (guild)
        {
            val schematic = (Schematic("/assets/tok/schematics/GuildCastle")).setPosition(posX, posY, posZ).setSpeed(300)
            SchematicHandler.INSTANCE.addBuilding(schematic)
            worldProvider.guild = true
        }
        else
        {
            val registry = SchematicRegistry.INSTANCE

            if (!registry.isRegistered)
            {
                registry.registerAllBuildings(posX, posY, posZ)

                worldProvider.townPosX = posX
                worldProvider.townPosY = posY
                worldProvider.townPosZ = posZ
                worldProvider.markDirty()
            }

            val schematic = registry.getSchematic(id).setSpeed(300)
            SchematicHandler.INSTANCE.addBuilding(schematic)
            Buildings.INSTANCE.setBuildingTrue(id)
        }
    }
}